### PR TITLE
Add "allowWritesOnUiThread" and "allowReadsOnUiThread" to RealmConfiguration

### PR DESCRIPTION
**Pull request by edualonso**
Added `allowWritesOnUiThread` to `RealmConfig` (and to both normal and sync builders). From now on **blocking transactions cannot be run from the UI to follow Google's guidelines** - see https://github.com/realm/realm-java/issues/2976#issuecomment-697327787. Calls to `executeTransaction` from the UI thread will yield a `RealmException` whereas users opting in will receive a Logcat warning. This new setting allows opting in for writes launched from the UI thread.

Added `allowReadsOnUiThread`to `RealmConfig` (and to both normal and sync builders). Queries have always been allowed to run on the UI thread and still are, but now there is an explicit setting to opt out. This setting behaves similarly to the one above.

Added `DynamicRealm.executeTransactionAsync` as a by-product of the aforementioned additions.

In short:

- It will not be possible to call `executeTransaction` (both for normal and dynamic Realms) from the UI by default - it will be possible to opt out with `allowWritesOnUiThread` but we do not recommend doing so.
- It will be possible to opt out of allowing launching queries (i.e. `findAll`, `findFirst`, `minimumDate`, `maximumDate`, `count`, `sum`, `max`, `min`, `average` and `averageDecimal128`) from the UI thread with `allowReadsOnUiThread`, though it will be allowed by default, though we do not recommend doing so.
- Calling `BaseRealm.refresh` from the UI thread will throw a `RealmException` if `allowReadsOnUiThread` has been set to false.

### CODE REVIEW

**Comment by edualonso** [comment]
Missing CHANGELOG updates, will do that later.

File: realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java [hunk 1]
```diff
@@ -1411,6 +1415,76 @@ public void onChange(RealmResults<AllTypes> results) {
...
+        assertEquals(1, results.size());
+        assertNotNull(results.first());
+        assertEquals("Snuffles", Objects.requireNonNull(results.first()).getName());
+
+        realm.close();
+    }
+
+    @Test
+    @UiThreadTest
+    public void executeTransactionAsync_throwsWhenRunningOnMainThread() {
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java:1451]: There is some misunderstanding here. Async transactions don't run on the UI thread, they are only started there, but actually run in a background thread, so this test and implementation is wrong. The only thing that should be disallowed is `executeTransaction()` (which runs on the same thread).

File: realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java [hunk 2]
```diff
@@ -1411,6 +1415,76 @@ public void onChange(RealmResults<AllTypes> results) {
         });
     }
 
+    @Test
+    @UiThreadTest
+    public void executeTransactionAsync_mainThreadQueriesAllowed() {
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java:1420]: Shouldn't this be named `executeTransaction_mainThreadWritesAllowed()`?

**Review by cmelchior — CHANGES_REQUESTED** [review]
There seem to be some misunderstanding about what to disallow. Async writes are fine, synchronous writes are not.

Also, any thoughts on https://github.com/realm/realm-java/issues/2976#issuecomment-696639590 (allowQueriesOnUiThread)?

**Comment by nirinchev** [comment]
What is the motivation for disallowing writes on the UI thread? Is it to protect developers from accidentally blocking the main thread if a background transaction is running or is there a legitimate consistency problem?

**Comment by cmelchior** [comment]
It is mostly to prevent people from shooting themselves in the foot. After having reviewed a few apps with Sync, it seems to be a common problem that people do writes on the UI thread. With synced Realms there is a fair chance that will lead to the app crashing with "App Not Responding" at some point.

For local Realm, it will probably mostly lead to dropped frames which is also bad.

But there are no consistency problems either way.

**Comment by edualonso** [comment]
It seems I had a neural path meltdown or something like that... I was totally oblivious and was on autopilot when I started writing the solution for executeTransactionAsyn. Of course async ones are alright 🤦‍♂️

**Comment by nhachicha** [comment]
Have we considered allowing the writes on UI threads by default (permissive) and trigger the exception behaviour if we opt-in (`allowWritesOnUiThread=false` or `forbidWritesOnUiThread`)?

**Comment by edualonso** [comment]
> Have we considered allowing the writes on UI threads by default (permissive) and trigger the exception behaviour if we opt-in (`allowWritesOnUiThread=false` or `forbidWritesOnUiThread`)?

That was the initial suggestion, but given v10 is a major release we decided to restrict them outright. We also decided to use "positive" words for the methods instead of "negative" ones.

**Comment by nirinchev** [comment]
Are we convinced `false` is the right default for that config option though? While it may be the right thing to do for production apps using sync, it will make it more cumbersome for people to get started and explore the SDK.

**Comment by cmelchior** [comment]
> Are we convinced false is the right default for that config option though? While it may be the right thing to do for production apps using sync, it will make it more cumbersome for people to get started and explore the SDK.

Yes. There are very strong recommendations for not doing I/O in the Android ecosystem so we should point people in the right direction already from the start. E.g. Room has the same default setting. And with the correct error message, it is easy to point people to the right solution.

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java [hunk 3]
```diff
@@ -82,6 +84,15 @@
      * to run it.
      */
     static <E extends RealmModel> RealmQuery<E> createQuery(Realm realm, Class<E> clazz) {
+        // Warn on query being executed on UI thread if isAllowQueriesOnUiThread is set to true, throw otherwise
+        if (realm.getSharedRealm().capabilities.isMainThread()) {
+            if (realm.getConfiguration().isAllowQueriesOnUiThread()) {
+                RealmLog.warn("It is not recommended to run queries on the UI thread as it may lead to a drop of frames or ANRs. Please consider doing so from another thread instead.");
+            } else {
+                throw new RealmException("There exists an opt-out for running queries on the UI thread. By default Realm allows queries from the UI thread. You can alternatively opt in using 'RealmConfiguration.allowQueriesOnUiThread'.");
```
Comment (edualonso) [inline realm/realm-library/src/main/java/io/realm/RealmQuery.java:92]:
```suggestion
                throw new RealmException("There exists an opt-out for running queries on the UI thread. By default Realm allows queries from the UI thread. You can alternatively opt in by using 'RealmConfiguration.allowQueriesOnUiThread'.");
```

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 4]
```diff
@@ -1525,6 +1525,15 @@ public void executeTransaction(Transaction transaction) {
             throw new IllegalArgumentException("Transaction should not be null");
         }
 
+        // Warn on transaction being executed on UI thread if allowWritesOnUiThread is set to true, throw otherwise
+        if (sharedRealm.capabilities.isMainThread()) {
+            if (getConfiguration().isAllowWritesOnUiThread()) {
+                RealmLog.warn("It is not recommended to run transactions on the UI thread as it may lead to a drop of frames or ANRs. Please consider using 'executeTransactionAsync' instead.");
+            } else {
+                throw new RealmException("Running transactions on the UI thread is disabled by default. You can opt in using 'RealmConfiguration.allowWritesOnUiThread'.");
```
Comment (edualonso) [inline realm/realm-library/src/main/java/io/realm/Realm.java:1533]:
```suggestion
                throw new RealmException("Running transactions on the UI thread is disabled by default. You can opt in by using 'RealmConfiguration.allowWritesOnUiThread'.");
```

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 5]
```diff
@@ -1525,6 +1525,15 @@ public void executeTransaction(Transaction transaction) {
             throw new IllegalArgumentException("Transaction should not be null");
         }
 
+        // Warn on transaction being executed on UI thread if allowWritesOnUiThread is set to true, throw otherwise
+        if (sharedRealm.capabilities.isMainThread()) {
+            if (getConfiguration().isAllowWritesOnUiThread()) {
+                RealmLog.warn("It is not recommended to run transactions on the UI thread as it may lead to a drop of frames or ANRs. Please consider using 'executeTransactionAsync' instead.");
+            } else {
+                throw new RealmException("Running transactions on the UI thread is disabled by default. You can opt in by using 'RealmConfiguration.allowWritesOnUiThread'.");
```
Comment (rorbech) [inline realm/realm-library/src/main/java/io/realm/Realm.java:1533]: This is not documented in `@throws`. The documentation could also point to `executeTransactionAsync` to guide users around this.

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java [hunk 6]
```diff
@@ -82,6 +84,15 @@
      * to run it.
      */
     static <E extends RealmModel> RealmQuery<E> createQuery(Realm realm, Class<E> clazz) {
+        // Warn on query being executed on UI thread if isAllowQueriesOnUiThread is set to true, throw otherwise
+        if (realm.getSharedRealm().capabilities.isMainThread()) {
+            if (realm.getConfiguration().isAllowQueriesOnUiThread()) {
+                RealmLog.warn("It is not recommended to run queries on the UI thread as it may lead to a drop of frames or ANRs. Please consider doing so from another thread instead.");
+            } else {
+                throw new RealmException("There exists an opt-out for running queries on the UI thread. By default Realm allows queries from the UI thread. You can alternatively opt in by using 'RealmConfiguration.allowQueriesOnUiThread'.");
```
Comment (rorbech) [inline realm/realm-library/src/main/java/io/realm/RealmQuery.java:92]: I guess we don't explicitly document throwing RealmExceptions from all the queries, but maybe some class level documentation about this pointing to the async variants?

File: realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java [hunk 7]
```diff
@@ -1411,6 +1415,101 @@ public void onChange(RealmResults<AllTypes> results) {
...
+                }
+            });
+            fail("It is not allowed to run blocking transactions on the UI thread by default.");
+        } catch (RealmException e) {
+            assertTrue(Objects.requireNonNull(e.getMessage()).contains("allowWritesOnUiThread"));
+        }
+    }
+
+    @Test
+    public void executeTransaction_runsOnNonUiThreadRegardlessOfSetting() {
```
Comment (rorbech) [inline realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java:1462]: Regardless of setting? Maybe set it explicitly to false and verify that it does not break and test default values in configuration tests.

File: realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java [hunk 8]
```diff
@@ -1411,6 +1415,101 @@ public void onChange(RealmResults<AllTypes> results) {
...
+            public void execute(Realm realm) {
+                // no-op
+            }
+        });
+    }
+
+    @Test
+    @UiThreadTest
+    public void query_runOnMainThreadAllowed() {
+        RealmConfiguration configuration = configFactory.createConfigurationBuilder()
```
Comment (rorbech) [inline realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java:1478]: Again, maybe add explicit `.allowQueriesOnUiThread(true)` and add configuration tests verifying default values.

**Review by rorbech — APPROVED** [review]
Looks good. Just minor test and documentation comments.

File: realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java (hunk 7 above)
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java:1462]: You are only testing one setting in this test. Shouldn't you manually create two different configurations? One with `allowWritesOnUiThread(true)` and one with `allowWritesOnUiThread(false)`?

File: realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java [hunk 9]
```diff
@@ -1411,6 +1415,101 @@ public void onChange(RealmResults<AllTypes> results) {
...
+        try (Realm realm = Realm.getInstance(configuration)) {
+            realm.where(Dog.class).findAll();
+            fail("This test specifies queries are not allowed to run on the UI thread, so something went awry.");
+        } catch (RealmException e) {
+            assertTrue(Objects.requireNonNull(e.getMessage()).contains("allowQueriesOnUiThread"));
+        }
+    }
+
+    @Test
+    public void query_runOnAnyThreadAllowed() {
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java:1503]: Shouldn't this be `query_runOnNonUiThreadAllowed()`?

File: realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java [hunk 10]
```diff
@@ -1411,6 +1415,101 @@ public void onChange(RealmResults<AllTypes> results) {
...
+                .build();
+
+        Realm realm = Realm.getInstance(configuration);
+        realm.where(Dog.class).findAll();
+        realm.close();
+    }
+
+    @Test
+    @UiThreadTest
+    public void query_runOnMainThreadThrows() {
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java:1488]: With this setting, you also need to check and throw if `Realm.refresh()` is being called, because that will convert async queries to sync queries which would violate this restriction.

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 11]
```diff
@@ -1525,6 +1525,15 @@ public void executeTransaction(Transaction transaction) {
             throw new IllegalArgumentException("Transaction should not be null");
         }
 
+        // Warn on transaction being executed on UI thread if allowWritesOnUiThread is set to true, throw otherwise
+        if (sharedRealm.capabilities.isMainThread()) {
+            if (getConfiguration().isAllowWritesOnUiThread()) {
+                RealmLog.warn("It is not recommended to run transactions on the UI thread as it may lead to a drop of frames or ANRs. Please consider using 'executeTransactionAsync' instead.");
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/Realm.java:1531]: Not sure we should log this as a warning if enabled. It is only enabled if you manually opt into it, so hopefully, people are already aware of the tradeoff and the warn level will most likely become visible to users since that is the default setting.

File: realm/realm-library/src/main/java/io/realm/Realm.java (hunk 5 above)
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/Realm.java:1533]:
```suggestion
                throw new RealmException("Running transactions on the UI thread are disabled by default. You can opt in by using 'RealmConfiguration.allowWritesOnUiThread', or avoid it by using 'Realm.executeTransactionAsync()'.");
```

File: realm/realm-library/src/main/java/io/realm/RealmConfiguration.java [hunk 12]
```diff
@@ -797,6 +829,28 @@ final Builder schema(Class<? extends RealmModel> firstClass, Class<? extends Rea
             return this;
         }
 
+        /**
+         * Sets whether or not users are allowed to perform calls to {@link Realm#executeTransaction} from the UI thread.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmConfiguration.java:833]:
```suggestion
         * Sets whether or not you are allowed to perform calls to {@link Realm#executeTransaction} from the UI thread. 
```

File: realm/realm-library/src/main/java/io/realm/RealmConfiguration.java [hunk 13]
```diff
@@ -797,6 +829,28 @@ final Builder schema(Class<? extends RealmModel> firstClass, Class<? extends Rea
...
+         * <p>
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmConfiguration.java:834]: The comment you had here is actually pretty great at describing the issue and work-around. I would add that description to this doc as well: https://github.com/realm/realm-java/pull/7123/files#diff-b77f99deee1ca52ca3dc3e47554e4e50R1531

File: realm/realm-library/src/main/java/io/realm/RealmConfiguration.java [hunk 14]
```diff
@@ -797,6 +829,28 @@ final Builder schema(Class<? extends RealmModel> firstClass, Class<? extends Rea
...
+         * with this method.</b>
+         */
+        public Builder allowWritesOnUiThread(boolean allowWritesOnUiThread) {
+            this.allowWritesOnUiThread = allowWritesOnUiThread;
+            return this;
+        }
+
+        /**
+         * Sets whether or not RealmQueries are allowed from the UI thread.
+         * <p>
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmConfiguration.java:845]: I would also add a line about being cautious about doing this and using `REalmQuery.findAllAsync()` is the preferred approach.

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/sync/SyncConfiguration.java [hunk 15]
```diff
@@ -1012,6 +1019,28 @@ public Builder maxNumberOfActiveVersions(long number) {
             return this;
         }
 
+        /**
+         * Sets whether or not users are allowed to perform calls to {@link Realm#executeTransaction} from the UI thread.
+         * <p>
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/mongodb/sync/SyncConfiguration.java:1024]: Any changes to the Javadoc in `RealmConfiguration.Builder()` should be reflected here.

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/sync/SyncConfiguration.java [hunk 16]
```diff
@@ -1012,6 +1019,28 @@ public Builder maxNumberOfActiveVersions(long number) {
...
+         * with this method.</b>
+         */
+        public Builder allowWritesOnUiThread(boolean allowWritesOnUiThread) {
+            this.allowWritesOnUiThread = allowWritesOnUiThread;
+            return this;
+        }
+
+        /**
+         * Sets whether or not RealmQueries are allowed from the UI thread.
+         * <p>
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/mongodb/sync/SyncConfiguration.java:1035]: Same

File: realm/realm-library/src/main/java/io/realm/RealmConfiguration.java [hunk 17]
```diff
@@ -288,6 +292,30 @@ public long getMaxNumberOfActiveVersions() {
...
+     * {@link Builder#allowWritesOnUiThread(boolean)} or its Realm Sync builder counterpart.</b>
+     *
+     * @return whether or not write operations are allowed to be run from the UI thread.
+     */
+    public boolean isAllowWritesOnUiThread() {
+        return allowWritesOnUiThread;
+    }
+
+    /**
+     * Returns whether {@code RealmQueries} are allowed from the UI thread.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmConfiguration.java:308]:
```suggestion
     * Returns whether {@link RealmQuery} are allowed from the UI thread.
```

**Review by cmelchior — CHANGES_REQUESTED** [review]

**Review by cmelchior — CHANGES_REQUESTED** [review]
Also missing a changelog entry under `#Breaking changes`

File: realm/realm-library/src/main/java/io/realm/Realm.java (hunk 11 above)
Comment (edualonso) [reply realm/realm-library/src/main/java/io/realm/Realm.java:1531]: What should we do then? User `RealmLog.info` instead?

File: realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java (hunk 10 above)
Comment (edualonso) [reply realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java:1488]: Yes, we discussed that yesterday. I'm still working on it 👍

File: realm/realm-library/src/main/java/io/realm/Realm.java (hunk 11 above)
Comment (cmelchior) [reply realm/realm-library/src/main/java/io/realm/Realm.java:1531]: I would probably just remove it completely

File: CHANGELOG.md [hunk 18]
```diff
@@ -1,11 +1,20 @@
...
+
 ## 10.0.0-BETA.8 (2020-09-23)
 
 We no longer support Realm Cloud (legacy), but instead the new MongoDB Realm Cloud. MongoDB Realm is a serverless platform that enables developers to quickly build applications without having to set up server infrastructure. MongoDB Realm is built on top of MongoDB Atlas, automatically integrating the connection to your database.
 
 The old Realm Cloud legacy APIs have undergone significant refactoring. The new APIs are all located in the `io.realm.mongodb` package with `io.realm.mongodb.App` as the entry point.
 
 ### Fixed
-* [RealmApp] Logging in caused an `token contains an invalid number of segments` error. (Issue [#7117](https://github.com/realm/realm-java/issues/7117), since 10.0.0-BETA.7) 
+* [RealmApp] Logging in caused an `token contains an invalid number of segments` error. (Issue [#7117](https://github.com/realm/realm-java/issues/7117), since 10.0.0-BETA.7)
```
Comment (edualonso) [inline CHANGELOG.md:17]: I've no idea why this appears as a modification. I didn't change anything here 🤷

Comment (edualonso) [reply CHANGELOG.md:17]: I see. The whitespace at the end got removed automatically.

File: CHANGELOG.md [hunk 19]
```diff
@@ -1,11 +1,20 @@
+## 10.0.0-BETA.9 (2020-XX-XX)
+
+### Enhancements
+* Users can now opt out from allowing queries to be launched from the UI thread by using `RealmConfiguration.Builder.allowQueriesOnUiThread(false)`. Doing so will result in a `RealmException` when calling `RealmQuery.findAll()` or `RealmQuery.findFirst()`. Queries will still be allowed from any thread by default, although we recommend using a non-UI thread to launch queries, or, alternatively, using `RealmQuery.findAllAsync()` or `RealmQuery.findFirstAsync()`
```
Comment (cmelchior) [inline CHANGELOG.md:4]: We should also mention `RealmQuery.<count/min/max/average>` and missing `.` at the end.

File: realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java [hunk 20]
```diff
@@ -1411,6 +1415,121 @@ public void onChange(RealmResults<AllTypes> results) {
...
+        RealmConfiguration configuration = configFactory.createConfigurationBuilder()
+                .allowQueriesOnUiThread(false)
+                .build();
+
+        // Try-with-resources
+        try (Realm realm = Realm.getInstance(configuration)) {
+            realm.where(Dog.class).findAll();
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java:1498]: We should also test `count/min/max/average`

File: realm/realm-library/src/main/java/io/realm/BaseRealm.java [hunk 21]
```diff
@@ -177,13 +177,17 @@ public boolean isAutoRefresh() {
...
      * <p>
      * WARNING: Calling this on a thread with async queries will turn those queries into synchronous queries.
-     * In most cases it is better to use {@link RealmChangeListener}s to be notified about changes to the
-     * Realm on a given thread than it is to use this method.
+     * This means this method will throw a {@link RealmException} if
+     * {@link RealmConfiguration.Builder#allowQueriesOnUiThread(boolean)} was used with {@code true} to
+     * obtain a Realm instance. In most cases it is better to use {@link RealmChangeListener}s to be notified
+     * about changes to the Realm on a given thread than it is to use this method.
      *
      * @throws IllegalStateException if attempting to refresh from within a transaction.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/BaseRealm.java:185]: We should add the RealmException here as well.

File: realm/realm-library/src/main/java/io/realm/BaseRealm.java [hunk 22]
```diff
@@ -493,6 +497,30 @@ protected void checkIfValid() {
...
     }
 
+    /**
+     * Checks whether queries are allowed from the UI thread in the current RealmConfiguration.
+     */
+    protected void checkAllowQueriesOnUiThread() {
+        // Warn on query being executed on UI thread if isAllowQueriesOnUiThread is set to true, throw otherwise
+        if (getSharedRealm().capabilities.isMainThread()) {
+            if (!getConfiguration().isAllowQueriesOnUiThread()) {
+                throw new RealmException("There exists an opt-out for running queries on the UI thread. By default Realm allows queries from the UI thread. You can alternatively opt in by using 'RealmConfiguration.allowQueriesOnUiThread'.");
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/BaseRealm.java:507]:
```suggestion
                throw new RealmException("Queries on the UI thread have been disabled. They can be enabled by setting 'RealmConfiguration.Builder.allowQueriesOnUiThread(true)'.");
```

File: realm/realm-library/src/main/java/io/realm/BaseRealm.java [hunk 23]
```diff
@@ -493,6 +497,30 @@ protected void checkIfValid() {
...
+    }
+
+    /**
+     * Checks whether writes are allowed from the UI thread in the current RealmConfiguration.
+     */
+    protected void checkAllowWritesOnUiThread() {
+        // Warn on transaction being executed on UI thread if allowWritesOnUiThread is set to true, throw otherwise
+        if (getSharedRealm().capabilities.isMainThread()) {
+            if (!getConfiguration().isAllowWritesOnUiThread()) {
+                throw new RealmException("Running transactions on the UI thread is disabled by default. You can opt in by using 'RealmConfiguration.allowWritesOnUiThread' or avoid it by using 'Realm.executeTransactionAsync'.");
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/BaseRealm.java:519]:
```suggestion
                throw new RealmException("Running transactions on the UI thread have been disabled. They can be enabled by setting 'RealmConfiguration.Builder.allowQueriesOnUiThread(true)'.");
```

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 24]
```diff
@@ -1514,17 +1514,25 @@ public void removeAllChangeListeners() {
...
+     * <p>
+     *
+     * Calling this method from the UI thread will throw a {@link RealmException}. Doing so may result in a drop of frames
+     * or even ANRs. We recommend calling this method from a non-UI thread or using
+     * {@link #executeTransactionAsync(Transaction)} instead.
      *
      * @param transaction the {@link io.realm.Realm.Transaction} to execute.
      * @throws IllegalArgumentException if the {@code transaction} is {@code null}.
      * @throws RealmMigrationNeededException if the latest version contains incompatible schema changes.
+     * @throws RealmException if called from the UI thread, unless an explicit opt-in has been declared in the configuration.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/Realm.java:1526]: Maybe link to the option?

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 25]
```diff
@@ -1514,17 +1514,25 @@ public void removeAllChangeListeners() {
      * Executes a given transaction on the Realm. {@link #beginTransaction()} and {@link #commitTransaction()} will be
      * called automatically. If any exception is thrown during the transaction {@link #cancelTransaction()} will be
      * called instead of {@link #commitTransaction()}.
+     * <p>
+     *
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/Realm.java:1518]:
```suggestion
```

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java [hunk 26]
```diff
@@ -51,6 +52,13 @@
  * is required.
  * <p>
  * A RealmQuery cannot be passed between different threads.
+ * <p>
+ * Queries can be launched from any thread and results are obtained fairly quickly most of the times. However, launching
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmQuery.java:56]: This is not true, you can only launch the query from the Thread holding the Realm instance. Same for the other descriptions

**Review by cmelchior — CHANGES_REQUESTED** [review]
Nice 👍 , but we still seem to lack implementation and tests for DynamicRealm methods and RealmQuery.<count/max/min/sum/average/minimumDate/maximumDate>

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java (hunk 26 above)
Comment (edualonso) [reply realm/realm-library/src/main/java/io/realm/RealmQuery.java:56]: Yeah, it was a broad generalisation. I will change it.

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java [hunk 27]
```diff
@@ -51,6 +52,12 @@
  * is required.
  * <p>
  * A RealmQuery cannot be passed between different threads.
+ * <p>
+ * Results are obtained fairly quickly most of the times. However, launching heavy queries from the UI thread may result
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmQuery.java:56]: Maybe it is just me but the term "fairly quickly" feels like a cop out. Either just say "Quickly" or remove the line completely IMO

**Review by cmelchior — CHANGES_REQUESTED** [review]
Nice. Getting there. I mostly have some comments about the new Dynamic Async transaction tests that could be cleaned up.

File: realm/realm-library/src/androidTest/java/io/realm/QueryTests.java [hunk 28]
```diff
@@ -45,6 +47,8 @@
     public final ExpectedException thrown = ExpectedException.none();
     @Rule
     public final RunInLooperThread looperThread = new RunInLooperThread();
+    @Rule
+    public final UiThreadTestRule uiThreadTestRule = new UiThreadTestRule();
```
[the remark this answers is not in the document]
Comment (edualonso) [reply realm/realm-library/src/androidTest/java/io/realm/QueryTests.java:51]: Not in this class, but in a child class for the `@UiThreadTest` annotation. Perhaps it's better to add it in that one instead.

[the remark this answers is not in the document]
Comment (edualonso) [reply realm/realm-library/src/androidTest/java/io/realm/QueryTests.java:51]: I remember now. This class has also some rules, so I thought it might as well be used like the "rule entry point".

File: realm/realm-library/src/androidTest/java/io/realm/DynamicRealmAsyncQueryTests.kt [hunk 29]
```diff
@@ -0,0 +1,370 @@
...
+        realm.executeTransactionAsync(DynamicRealm.Transaction { transactionRealm ->
+            val owner = transactionRealm.createObject(Owner.CLASS_NAME)
+            owner.setString(Owner.FIELD_NAME, "Owner")
+        }, DynamicRealm.Transaction.OnSuccess {
+            assertTrue(realm.isClosed)
+            val newRealm = DynamicRealm.getInstance(config)
+            assertEquals(1, newRealm.where(Owner.CLASS_NAME).count())
+            assertEquals("Owner", newRealm.where(Owner.CLASS_NAME).findFirst()!!.getString(Owner.FIELD_NAME))
+
+            newRealm.close()
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/DynamicRealmAsyncQueryTests.kt:113]: I would do `DynamicRealm.getInstance(config).use { newRealm -> ... }` . Otherwise, if the assserts fail, this code is not called.

File: realm/realm-library/src/androidTest/java/io/realm/DynamicRealmAsyncQueryTests.kt [hunk 30]
```diff
@@ -0,0 +1,370 @@
...
+        val realm = DynamicRealm.getInstance(config)
+        val runtimeException = RuntimeException("Oh! What a Terrible Failure")
+
+        assertEquals(0, realm.where(Owner.CLASS_NAME).count())
+
+        realm.executeTransactionAsync({
+            throw runtimeException
+        }) { error ->
+            assertTrue(realm.isClosed)
+            val newRealm = DynamicRealm.getInstance(config)
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/DynamicRealmAsyncQueryTests.kt:151]: Same

**Review by cmelchior — APPROVED** [review]
Woop woop 🎉
